File: nuam-backend/calificaciones/validators.py
import re
from rest_framework import serializers

def validate_rut(value):
    """
    Validates a Chilean RUT.
    Expected format: XXXXXXXX-Y or XX.XXX.XXX-Y (will be cleaned)
    """
    if not value:
        return value

    # Clean format
    rut_clean = value.replace(".", "").replace("-", "").upper()
    
    if len(rut_clean) < 2:
        raise serializers.ValidationError("El RUT es demasiado corto.")

    body = rut_clean[:-1]
    dv = rut_clean[-1]

    # Validate body is numeric
    if not body.isdigit():
        raise serializers.ValidationError("El cuerpo del RUT debe ser numérico.")

    # Calculate DV
    suma = 0
    multiplo = 2
    
    for c in reversed(body):
        suma += int(c) * multiplo
        multiplo += 1
        if multiplo == 8:
            multiplo = 2
            
    res = 11 - (suma % 11)
    
    if res == 11:
        dv_calc = '0'
    elif res == 10:
        dv_calc = 'K'
    else:
        dv_calc = str(res)
        
    if dv != dv_calc:
        raise serializers.ValidationError("El RUT no es válido (dígito verificador incorrecto).")
        
    return value

def validate_positive(value):
    """
    Validates that the value is positive (> 0).
    """
    if value is not None and value < 0:
        raise serializers.ValidationError("El valor debe ser positivo.")
    return value

File: nuam-frontend/src/utils/validators.js
export const cleanRut = (rut) => {
    return typeof rut === 'string' ? rut.replace(/^0+|[^0-9kK]+/g, '').toUpperCase() : '';
};

export const validateRut = (rut) => {
    if (!rut) return false;

    const clean = cleanRut(rut);
    if (clean.length < 2) return false;

    const body = clean.slice(0, -1);
    const dv = clean.slice(-1);

    if (!/^\d+$/.test(body)) return false;

    let suma = 0;
    let multiplo = 2;

    for (let i = body.length - 1; i >= 0; i--) {
        suma += parseInt(body.charAt(i)) * multiplo;
        multiplo = multiplo < 7 ? multiplo + 1 : 2;
    }

    const dvCalc = 11 - (suma % 11);
    const dvEsperado = dvCalc === 11 ? '0' : dvCalc === 10 ? 'K' : dvCalc.toString();

    return dv === dvEsperado;
};

export const formatRut = (rut) => {
    if (!rut) return '';

    const clean = cleanRut(rut);
    if (clean.length <= 1) return clean;

    const body = clean.slice(0, -1);
    const dv = clean.slice(-1);

    let formattedBody = '';
    for (let i = body.length - 1, j = 0; i >= 0; i--, j++) {
        formattedBody = body.charAt(i) + (j > 0 && j % 3 === 0 ? '.' : '') + formattedBody;
    }

    return `${formattedBody}-${dv}`;
};
